src/target/riscv/riscv.c

```diff
@@ -1601,9 +1602,17 @@ static int riscv_trigger_detect_hit_bits(struct target *target, int64_t *unique_
 				break;
 			case CSR_TDATA1_TYPE_MCONTROL:
 				hit_mask = CSR_MCONTROL_HIT;
+				if (get_field(tdata1, CSR_MCONTROL_TIMING) == CSR_MCONTROL_TIMING_BEFORE)
+					r->need_single_step = true;
```

Comment:
Here it is only known that there is a trigger with type `mcontrol`. It is not yet known that this trigger is the one that was hit . To know that one needs to check `tdata1 & hit_mask`.
```suggestion
					if (tdata1 & hit_mask)
						r->need_single_step = true;
```

Also, the spec states [5.7.11. Match Control (mcontrol, at 0x7a1)] the `hit` field description:
> If this bit is implemented then it must become set when this trigger fires and may become set when this trigger matches.

Therefore one can be sure that this trigger with type `mcontrol` has fired only if it's the only trigger that has it's `hit` field set to a non-zero value and all other triggers support `hit` fields. The last part is the most problematic -- OpenOCD does not gather information of `hit` field support at the moment. Though it can be done by changing how the triggers are set: `hit` fields are WARL and therefore `hit` field support can be discovered by writing to it when setting a trigger.

There is no such issue for `mcontrol6`triggers since [5.7.12. Match Control Type 6 (mcontrol6, at 0x7a1)], `hit0` field description:
> The TM updates this field when the trigger fires.

I'd suggest adding a `TODO`/filing an issue about it and only setting `r->need_single_step = true` for `mcontrol6` triggers for now.

Reply:
addressed, and checking  `mcontrol.hit` not implemented scenario

Reply:
Though it can be done by changing how the triggers are set: hit fields are WARL and therefore hit field support can be discovered by writing to it when setting a trigger.
——————————
now，this patch using this solution and checking `hit` bits whether implemented